File: solutions/python/bank-account/1/bank_account.py
class BankAccount:
    def __init__(self):
        self.is_open = False

    def get_balance(self):
        if self.is_open:
            return self.balance
        else:
            raise ValueError("account not open")

    def open(self):
        if not self.is_open:
            self.is_open = True
            self.balance = 0
        else:
            raise ValueError("account already open")

    def deposit(self, amount):
        if self.is_open:
            if amount < 0:
                raise ValueError("amount must be greater than 0")
            self.balance += amount
        else:
            raise ValueError("account not open")

    def withdraw(self, amount):
        if self.is_open:
            if amount < 0:
                raise ValueError("amount must be greater than 0")
            if amount > self.balance:
                raise ValueError("amount must be less than balance")

            self.balance -= amount
        else:
            raise ValueError("account not open")

    def close(self):
        if self.is_open:
            self.is_open = False
        else:
            raise ValueError("account not open")

File: solutions/javascript/bank-account/1/bank-account.js
export class BankAccount {
  constructor() {
    this._open = false;
    this._balance = 0;
  }

  open() {
    if (!this._open) {
      this._open = true;
      this._balance = 0;
    } else {
      throw new ValueError();
    }
  }

  close() {
    if (this._open) {
      this._open = false;
    } else {
      throw new ValueError();
    }
  }

  deposit(amount) {
    if (this._open && amount >= 0) {
      this._balance += amount;
    } else {
      throw new ValueError();
    }
  }

  withdraw(amount) {
    if (this._open && amount <= this._balance && amount >= 0) {
      this._balance -= amount;
    } else {
      throw new ValueError();
    }
  }

  get balance() {
    if (this._open) {
      return this._balance;
    } else {
      throw new ValueError();
    }
  }
}

export class ValueError extends Error {
  constructor() {
    super("Bank account error");
  }
}
